biz.aQute.bndlib/src/aQute/bnd/stream/EntryPipeline.java

```diff
@@ -0,0 +1,312 @@
+	}
+
+	@Override
+	public <R, S> MapStream<R, S> flatMap(
+		BiFunction<? super K, ? super V, ? extends MapStream<? extends R, ? extends S>> mapper) {
+		requireNonNull(mapper);
+		return new EntryPipeline<>(entries().flatMap(e -> mapper.apply(e.getKey(), e.getValue())
+			.entries()));
+	}
+
+	@Override
+	public <R> Stream<R> flatMapToObj(BiFunction<? super K, ? super V, ? extends Stream<? extends R>> mapper) {
+		requireNonNull(mapper);
+		return entries().flatMap(e -> mapper.apply(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public IntStream flatMapToInt(BiFunction<? super K, ? super V, ? extends IntStream> mapper) {
+		requireNonNull(mapper);
+		return entries().flatMapToInt(e -> mapper.apply(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public LongStream flatMapToLong(BiFunction<? super K, ? super V, ? extends LongStream> mapper) {
+		requireNonNull(mapper);
+		return entries().flatMapToLong(e -> mapper.apply(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public DoubleStream flatMapToDouble(BiFunction<? super K, ? super V, ? extends DoubleStream> mapper) {
+		requireNonNull(mapper);
+		return entries().flatMapToDouble(e -> mapper.apply(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public MapStream<K, V> peek(BiConsumer<? super K, ? super V> peek) {
+		requireNonNull(peek);
+		return new EntryPipeline<>(entries().peek(e -> peek.accept(e.getKey(), e.getValue())));
+	}
+
+	@Override
+	public MapStream<K, V> sortedByKey() {
+		@SuppressWarnings("unchecked")
+		Comparator<K> comparator = (Comparator<K>) Comparator.naturalOrder();
+		return sortedByKey(comparator);
+	}
+
+	@Override
+	public MapStream<K, V> sortedByKey(Comparator<? super K> comparator) {
+		return new EntryPipeline<>(entries().sorted(Entry.comparingByKey(comparator)));
+	}
+
+	@Override
+	public MapStream<K, V> sortedByValue() {
+		@SuppressWarnings("unchecked")
+		Comparator<V> comparator = (Comparator<V>) Comparator.naturalOrder();
+		return sortedByValue(comparator);
+	}
+
+	@Override
+	public MapStream<K, V> sortedByValue(Comparator<? super V> comparator) {
+		return new EntryPipeline<>(entries().sorted(Entry.comparingByValue(comparator)));
+	}
+
+	@Override
+	public MapStream<K, V> limit(long maxSize) {
+		return new EntryPipeline<>(entries().limit(maxSize));
+	}
+
+	@Override
+	public MapStream<K, V> skip(long n) {
+		return new EntryPipeline<>(entries().skip(n));
+	}
+
+	@Override
+	public void forEach(BiConsumer<? super K, ? super V> consumer) {
+		requireNonNull(consumer);
+		entries().forEach(e -> consumer.accept(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public void forEachOrdered(BiConsumer<? super K, ? super V> consumer) {
+		requireNonNull(consumer);
+		entries().forEachOrdered(e -> consumer.accept(e.getKey(), e.getValue()));
+	}
+
+	@Override
+	public long count() {
+		return entries().count();
+	}
+
+	@Override
+	public boolean anyMatch(BiPredicate<? super K, ? super V> predicate) {
```

Comment:
Not sure it's really necessary but for completeness should we have `anyMatchByKey|Value` and so forth for `allMatch`, `noneMatch`?

Reply:
I don’t think `anyMatchByKey`, etc. is useful since you can just ignore the key or value argument. `mapKey` is helpful since you avoid having to make an `Entry` for the result. These `any`/`all`/`none` method just return boolean. 

A reason to add them would be to use an existing `Predicate` instead of having to make a `BiPredicate`. Do you think this makes adding them worthwhile?

Reply:
The more I think about it, perhaps it is a good idea to add them.

Reply:
Actually, I now recall why these methods are not necessary. Since the match methods are terminals which return a boolean, there is not need to preserve the entries in the stream like we must do for filterKey/filterValue. To match by key call the match method on `keys()` and to match by value call the match method on `values()`. For example `mapStream.keys().anyMatch(predicate)`. This is basically what an `anyMatchKey` method would do anyway. Similarly if you wanted to find the max key, you would do `mapStream.keys().max(comparator)`.

Reply:
Ok makes sense!